NIFI-10403 Add processor supporting the new BigQuery Write API

bejancsaba opened the pull request:
<!-- Licensed to the Apache Software Foundation (ASF) under one or more -->
<!-- contributor license agreements.  See the NOTICE file distributed with -->
<!-- this work for additional information regarding copyright ownership. -->
<!-- The ASF licenses this file to You under the Apache License, Version 2.0 -->
<!-- (the "License"); you may not use this file except in compliance with -->
<!-- the License.  You may obtain a copy of the License at -->
<!--     http://www.apache.org/licenses/LICENSE-2.0 -->
<!-- Unless required by applicable law or agreed to in writing, software -->
<!-- distributed under the License is distributed on an "AS IS" BASIS, -->
<!-- WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. -->
<!-- See the License for the specific language governing permissions and -->
<!-- limitations under the License. -->

# Summary

[NIFI-10403](https://issues.apache.org/jira/browse/NIFI-10403)

# Tracking

Please complete the following tracking steps prior to pull request creation.

### Issue Tracking

- [x] [Apache NiFi Jira](https://issues.apache.org/jira/browse/NIFI) issue created

### Pull Request Tracking

- [x] Pull Request title starts with Apache NiFi Jira issue number, such as `NIFI-00000`
- [x] Pull Request commit message starts with Apache NiFi Jira issue number, as such `NIFI-00000`

### Pull Request Formatting

- [x] Pull Request based on current revision of the `main` branch
- [x] Pull Request refers to a feature branch with one commit containing changes

# Verification

Please indicate the verification steps performed prior to pull request creation.

### Build

- [x] Build completed using `mvn clean install -P contrib-check`
  - [x] JDK 8
  - [ ] JDK 11
  - [ ] JDK 17

### Licensing

- [x] New dependencies are compatible with the [Apache License 2.0](https://apache.org/licenses/LICENSE-2.0) according to the [License Policy](https://www.apache.org/legal/resolved.html)
- [ ] New dependencies are documented in applicable `LICENSE` and `NOTICE` files

### Documentation

- [x] Documentation formatting appears as expected in rendered files

ferencerdei commented on nifi-nar-bundles/nifi-gcp-bundle/nifi-gcp-processors/src/main/java/org/apache/nifi/processors/gcp/bigquery/PutBigQueryStreaming.java:
```diff
@@ -67,6 +67,8 @@
+ * @deprecated use {@link PutBigQuery} instead which uses the Write API
  */
 @InputRequirement(InputRequirement.Requirement.INPUT_REQUIRED)
 @Tags({ "google", "google cloud", "bq", "gcp", "bigquery", "record" })
```
could you add the deprecation notice into the CapabilityDescription so users will be notified as well about alternatives

bejancsaba replied:
Done

ferencerdei reviewed (changes requested):
Thanks for the improved GCP processor. I have some minor comments / suggestions.

bejancsaba commented on nifi-nar-bundles/nifi-gcp-bundle/nifi-gcp-processors/src/main/java/org/apache/nifi/processors/gcp/bigquery/PutBigQuery.java:
```diff
@@ -0,0 +1,470 @@
+    }
+
+    private static Map<String, Object> convertMapRecord(Map<String, Object> map) {
+        Map<String, Object> result = new HashMap<String, Object>();
```
Applied

bejancsaba commented on nifi-nar-bundles/nifi-gcp-bundle/nifi-gcp-processors/src/main/java/org/apache/nifi/processors/gcp/bigquery/PutBigQuery.java:
```diff
@@ -0,0 +1,470 @@
+            } else if (obj instanceof Object[]
+                && ((Object[]) obj).length > 0
+                && ((Object[]) obj)[0] instanceof MapRecord) {
+                List<Map<String, Object>> lmapr = new ArrayList<Map<String, Object>>();
```
Applied

bejancsaba commented on nifi-nar-bundles/nifi-gcp-bundle/nifi-gcp-processors/src/main/java/org/apache/nifi/processors/gcp/bigquery/PutBigQuery.java:
```diff
@@ -0,0 +1,470 @@
+
+            Map<String, String> attributes = new HashMap<>();
+            attributes.put(BigQueryAttributes.JOB_NB_RECORDS_ATTR, Integer.toString(recordNumWritten));
+            flowFile = session.putAllAttributes(flowFile, attributes);
```
Thanks I wasn't aware, applied

bejancsaba commented on nifi-nar-bundles/nifi-gcp-bundle/nifi-gcp-processors/src/main/java/org/apache/nifi/processors/gcp/bigquery/PutBigQuery.java:
```diff
@@ -0,0 +1,470 @@
+        inflightRequestCount.register();
+    }
+
+    private void finalize(ProcessSession session, FlowFile flowFile, StreamWriter streamWriter, String streamName, String parentTable) {
```
Renamed to finishProcessing

bejancsaba commented on nifi-nar-bundles/nifi-gcp-bundle/nifi-gcp-processors/src/main/java/org/apache/nifi/processors/gcp/bigquery/PutBigQuery.java:
```diff
@@ -0,0 +1,470 @@
+            }
+
+            error.compareAndSet(null, Optional.ofNullable(Exceptions.toStorageException(throwable))
+                .map(storageException -> (RuntimeException) storageException)
```
Changed

bejancsaba commented on nifi-nar-bundles/nifi-gcp-bundle/nifi-gcp-processors/src/main/java/org/apache/nifi/processors/gcp/bigquery/PutBigQuery.java:
```diff
@@ -0,0 +1,470 @@
+            getLogger().error("Error occurred in the stream: ", error.get());
+            Map<String, String> attributes = new HashMap<>();
+            attributes.put(BigQueryAttributes.JOB_NB_RECORDS_ATTR, isBatch() ? "0" : String.valueOf(appendSuccessCount.get() * recordBatchCount));
+            flowFile = session.putAllAttributes(flowFile, attributes);
```
Done

bejancsaba commented on nifi-nar-bundles/nifi-gcp-bundle/nifi-gcp-processors/src/main/java/org/apache/nifi/processors/gcp/bigquery/PutBigQuery.java:
```diff
@@ -0,0 +1,470 @@
+                // If the response does not have a commit time, it means the commit operation failed.
+                if (!commitResponse.hasCommitTime()) {
+                    for (StorageError err : commitResponse.getStreamErrorsList()) {
+                        getLogger().error("Commit operation error: ", err.getErrorMessage());
```
As there could be multiple errors on the batch it could lead to confusion if we want to aggregate those so I would leave it as is.

bejancsaba commented:
> Thanks for the improved GCP processor. I have some minor comments / suggestions.

@ferencerdei Thanks for taking a look, I addressed your comments.

ferencerdei reviewed (approved):
Thanks for the update +1 from my side

arpadboda reviewed (commented):
Added two minor comments, mostly looks good to me with the latest changes.

exceptionfactory commented on nifi-nar-bundles/nifi-gcp-bundle/nifi-gcp-processors/src/test/java/org/apache/nifi/processors/gcp/bigquery/PutBigQueryBatchIT.java:
```diff
@@ -32,6 +32,10 @@
 import java.nio.file.Files;
 import java.nio.file.Path;
 
+/**
+ * @deprecated {@link PutBigQueryIT} is covering the needed validation for the new class.
+ */
+@Deprecated
```
Deprecating the integration test is generally unnecessary. It would actually be worth removing the integration test, because it is not run as part of standard builds.

bejancsaba replied:
Wasn't sure what to do here. I agree with you I removed the test.

exceptionfactory commented on nifi-nar-bundles/nifi-gcp-bundle/nifi-gcp-processors/src/main/resources/docs/org.apache.nifi.processors.gcp.bigquery.PutBigQuery/additionalDetails.html:
```diff
@@ -0,0 +1,58 @@
+</p>
+
+<p>
+    You can find additional details on the official <a href="https://cloud.google.com/bigquery/docs/write-api">Write API site</a>
```
It is best to avoid the second-person `You` in documentation.
```suggestion
    The official <a href="https://cloud.google.com/bigquery/docs/write-api">Google Write API documentation</a> provides additional details.
```

bejancsaba replied:
Thanks, applied

exceptionfactory commented on nifi-nar-bundles/nifi-gcp-bundle/nifi-gcp-processors/src/test/java/org/apache/nifi/processors/gcp/bigquery/PutBigQueryIT.java:
```diff
@@ -0,0 +1,355 @@
+    }
+
+    @Test
+    public void PutBigQueryStreamingNoError() throws Exception {
```
Method names should follow standard Java conventions, beginning with a lowercase letter. In this case, recommend starting each method with `test`. It is also unnecessary to repeat the component name in the method.
```suggestion
    public void testStreamingNoError() throws Exception {
```

bejancsaba replied:
Applied the comment to all the tests in the class.

exceptionfactory commented on nifi-nar-bundles/nifi-gcp-bundle/nifi-gcp-processors/src/test/java/org/apache/nifi/processors/gcp/bigquery/PutBigQueryStreamingIT.java:
```diff
@@ -47,6 +47,10 @@
 import static org.junit.Assert.assertFalse;
 import static org.junit.Assert.assertTrue;
 
+/**
+ * @deprecated {@link PutBigQueryIT} is covering the needed validation for the new class.
+ */
+@Deprecated
```
As mentioned on the other integration test, recommend removing the class instead of deprecating it.

bejancsaba replied:
Similalrily to the other, removed the test

exceptionfactory commented on nifi-nar-bundles/nifi-gcp-bundle/nifi-gcp-processors/src/main/java/org/apache/nifi/processors/gcp/bigquery/PutBigQuery.java:
```diff
@@ -0,0 +1,457 @@
+
+    static final String STREAM = "STREAM";
+    static final String BATCH = "BATCH";
+    static final AllowableValue STREAM_TYPE = new AllowableValue(STREAM, STREAM, "Define streaming approach.");
```
Recommend adjusting the description to something along the lines of "Use streaming record handling strategy".
```suggestion
    static final AllowableValue STREAM_TYPE = new AllowableValue(STREAM, STREAM, "Use streaming record handling strategy");
```

bejancsaba replied:
Applied

exceptionfactory commented on nifi-nar-bundles/nifi-gcp-bundle/nifi-gcp-processors/src/main/java/org/apache/nifi/processors/gcp/bigquery/PutBigQuery.java:
```diff
@@ -0,0 +1,457 @@
+
+    @Override
+    public List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        List<PropertyDescriptor> descriptors = new ArrayList<>(super.getSupportedPropertyDescriptors());
+        descriptors.add(TRANSFER_TYPE);
+        descriptors.add(RECORD_READER);
+        descriptors.add(APPEND_RECORD_COUNT);
+        descriptors.add(SKIP_INVALID_ROWS);
+        descriptors.remove(IGNORE_UNKNOWN);
+
+        return Collections.unmodifiableList(descriptors);
```
The list of descriptors should generally be defined in a static initializer and returned in this method, to avoid unnecessary construction of the Collection for every invocation.

bejancsaba replied:
There are descriptors coming from the Abstract BigQuery and on top of that Abstract GCP class and those are not defined as static so without changing the shared logic for all GCP processors I can't apply this change. I checked the other GCP processors and they use the same approach as here. Are you ok with keeping this as is and if you think we can introduce enhancements for all GCP processors as part of separate hardening ticket?

exceptionfactory replied:
Thanks for the reply, I missed the super method call in the previous review, the current approach looks good.

exceptionfactory commented on nifi-nar-bundles/nifi-gcp-bundle/nifi-gcp-processors/src/main/java/org/apache/nifi/processors/gcp/bigquery/PutBigQuery.java:
```diff
@@ -0,0 +1,457 @@
+    }
+
+    @OnUnscheduled
+    public void onUnScheduled() {
```
```suggestion
    public void onUnscheduled() {
```

bejancsaba replied:
Chenged, thanks

exceptionfactory commented on nifi-nar-bundles/nifi-gcp-bundle/nifi-gcp-processors/src/main/java/org/apache/nifi/processors/gcp/bigquery/PutBigQuery.java:
```diff
@@ -0,0 +1,457 @@
+            protoDescriptor = BQTableSchemaToProtoDescriptor.convertBQTableSchemaToProtoDescriptor(writeStream.getTableSchema());
+            streamWriter = createStreamWriter(writeStream.getName(), protoDescriptor, getGoogleCredentials(context));
+        } catch (Descriptors.DescriptorValidationException | IOException e) {
+            getLogger().error("Failed to create Big Query Stream Writer for writing due to {}", e);
```
The exception should not be parameterized as a placeholder, it will be appended with a stack trace.
```suggestion
            getLogger().error("Failed to create Big Query Stream Writer for writing", e);
```

bejancsaba replied:
Applied

exceptionfactory commented on nifi-nar-bundles/nifi-gcp-bundle/nifi-gcp-processors/src/main/java/org/apache/nifi/processors/gcp/bigquery/PutBigQuery.java:
```diff
@@ -0,0 +1,457 @@
+
+            flowFile = session.putAttribute(flowFile, BigQueryAttributes.JOB_NB_RECORDS_ATTR, Integer.toString(recordNumWritten));
+        } catch (Exception e) {
+            getLogger().error(e.getMessage(), e);
```
A specific message should be logged, instead of repeating the exception `getMessage()`, which will appear in the stack trace.
```suggestion
            getLogger().error("Writing Records failed", e);
```

bejancsaba replied:
Applied

exceptionfactory commented on nifi-nar-bundles/nifi-gcp-bundle/nifi-gcp-processors/src/main/java/org/apache/nifi/processors/gcp/bigquery/PutBigQuery.java:
```diff
@@ -0,0 +1,457 @@
+            flowFile = session.putAttribute(flowFile, BigQueryAttributes.JOB_NB_RECORDS_ATTR, Integer.toString(recordNumWritten));
+        } catch (Exception e) {
+            getLogger().error(e.getMessage(), e);
+            error.set(new RuntimeException(e));
```
Recommend not wrapping the Exception in a `RuntimeException`:
```suggestion
            error.set(e);
```

bejancsaba replied:
Changed the logic here

exceptionfactory commented on nifi-nar-bundles/nifi-gcp-bundle/nifi-gcp-processors/src/main/java/org/apache/nifi/processors/gcp/bigquery/PutBigQuery.java:
```diff
@@ -0,0 +1,457 @@
+        try {
+            message = ProtoUtils.createMessage(descriptor, valueMap);
+        } catch (RuntimeException e) {
+            getLogger().error("Can't create message from input", e);
```
Recommend avoiding contractions like `Can't` in logs and spelling out the word instead:
```suggestion
            getLogger().error("Cannot convert record to message", e);
```

bejancsaba replied:
Changed

exceptionfactory commented on nifi-nar-bundles/nifi-gcp-bundle/nifi-gcp-processors/src/main/java/org/apache/nifi/processors/gcp/bigquery/PutBigQuery.java:
```diff
@@ -0,0 +1,457 @@
+
+        // Verify that no error occurred in the stream.
+        if (error.get() != null) {
+            getLogger().error("Error occurred in the stream: ", error.get());
```
The colon character should not be included.
```suggestion
            getLogger().error("Stream processing failed", error.get());
```

bejancsaba replied:
Removed

exceptionfactory commented on nifi-nar-bundles/nifi-gcp-bundle/nifi-gcp-processors/src/main/java/org/apache/nifi/processors/gcp/bigquery/PutBigQuery.java:
```diff
@@ -0,0 +1,457 @@
+                // If the response does not have a commit time, it means the commit operation failed.
+                if (!commitResponse.hasCommitTime()) {
+                    for (StorageError err : commitResponse.getStreamErrorsList()) {
+                        getLogger().error("Commit operation error: ", err.getErrorMessage());
```
The error should include a placeholder for the message. Does `StorageError` contain any other useful information that should be logged?
```suggestion
                        getLogger().error("Commit Storage Error: {}", err.getErrorMessage());
```

bejancsaba replied:
Updated and also added logging of the error code which could be useful for investigation. Thanks.

exceptionfactory commented on nifi-nar-bundles/nifi-gcp-bundle/nifi-gcp-processors/src/main/java/org/apache/nifi/processors/gcp/bigquery/PutBigQuery.java:
```diff
@@ -0,0 +1,457 @@
+        }
+
+        public void onSuccess(AppendRowsResponse response) {
+            getLogger().info("Append success with offset: " + appendContext.getOffset());
```
Message concatenation should be avoided for logs, placeholders should be used.
```suggestion
            getLogger().info("Append success with offset: {}", appendContext.getOffset());
```

bejancsaba replied:
Changed

exceptionfactory commented on nifi-nar-bundles/nifi-gcp-bundle/nifi-gcp-processors/src/main/java/org/apache/nifi/processors/gcp/bigquery/PutBigQuery.java:
```diff
@@ -0,0 +1,457 @@
+                    inflightRequestCount.arriveAndDeregister();
+                    return;
+                } catch (Exception e) {
+                    getLogger().error("Failed to retry append: ", e);
```
```suggestion
                    getLogger().error("Failed to retry append", e);
```

bejancsaba replied:
Removed

exceptionfactory commented on nifi-nar-bundles/nifi-gcp-bundle/nifi-gcp-processors/src/main/java/org/apache/nifi/processors/gcp/bigquery/PutBigQuery.java:
```diff
@@ -0,0 +1,457 @@
+                .map(RuntimeException.class::cast)
+                .orElse(new RuntimeException(throwable)));
+
+            getLogger().error("Failure during appending data: ", throwable);
```
```suggestion
            getLogger().error("Failure during appending data", throwable);
```

bejancsaba replied:
Removed

exceptionfactory commented on nifi-nar-bundles/nifi-gcp-bundle/nifi-gcp-processors/src/main/java/org/apache/nifi/processors/gcp/bigquery/PutBigQuery.java:
```diff
@@ -0,0 +1,457 @@
+        try {
+            client = BigQueryWriteClient.create(BigQueryWriteSettings.newBuilder().setCredentialsProvider(FixedCredentialsProvider.create(credentials)).build());
+        } catch (Exception e) {
+            throw new ProcessException("Failed to create Big Query Write Client for writing due to", e);
```
```suggestion
            throw new ProcessException("Failed to create Big Query Write Client for writing", e);
```

bejancsaba replied:
Removed

exceptionfactory commented on nifi-nar-bundles/nifi-gcp-bundle/nifi-gcp-processors/src/main/java/org/apache/nifi/processors/gcp/bigquery/PutBigQuery.java:
```diff
@@ -0,0 +1,457 @@
+                    lmapr.add(convertMapRecord(((MapRecord) mapr).toMap()));
+                }
+                result.put(key, lmapr);
+            } else if (obj instanceof Timestamp) {
+                // ZoneOffset.UTC time zone is necessary due to implicit time zone conversion in Record Readers from
+                // the local system time zone to the GMT time zone
```
This should no longer be happening, are you sure this is the case?

bejancsaba replied:
When I was working on the implementation I was struggling a lot Time / Timestamp it seemed that I'm always off with 1 or 2 hours when the data landed on the BigQuery side. I remember looking into what is happening and the RecordReader works with DateTypeUtils class. When the date/time is a long number (epoch) we are good however if it is a format string it uses the dateFormat to parse it, however for example the AbstractCsvRecordReader initialises the DateFormat as follwos
```
LAZY_TIME_FORMAT = () -> DataTypeUtils.getDateFormat(timeFormat);
```
and the DataTypeUtils.getDateFormat(timeFormat) is using the default timeZone "return getDateFormat(pattern, TimeZone.getDefault());" and I think that is where things go south but maybe I'm missing something I can look into this more deeply if needed.

exceptionfactory replied:
Thanks for the response. Not wanting to hold things up, but I think this is worth a bit more investigation. The `TimeZone.getDefault()` method returns the time zone according to the operating system and Java platform configuration, which is usually a local time zone, but could be UTC depending on the configuration. Timestamp strings without a zone offset should be handled using the default timezone. For that reason, performing this type of conversion in the processor itself should be avoided.

When timestamps are formatted as strings, they should include a timezone offset to guarantee the same behavior regardless of the system configuration. In absence of the timezone offset, it is best to fallback to the system default timezone, instead of attempting to make adjustments.

In older versions of NiFi, the DateTypeUtils had an internal setting that defaulted to GMT, but this created other complex problems in referencing components, so that behavior was moved.

With that background, the best approach is to avoid any automatic timezone conversion.

bejancsaba replied:
I looked into the whole thing, tested it manually and adjusted the input and the integration tests as well. It looks like I managed to make it work and verify the correctness. Date and time handling is always a challenge in this case when RecordReader, NiFi execution environment and Google Big Query are all involved with mapping between different types along the way (java, grpc, google bigquery storage...) it is not trivial for sure :) Thanks for looking into it I had challenges with keeping my test in sync with the existing ones but I think the existing was not working correctly so removing it made everything clearer.

exceptionfactory replied:
Thanks for making the adjustments to the date and time handling. I know some older components previously included workarounds related to DateTypeUtils, so this was a good opportunity to streamline the approach. The updates look good.

exceptionfactory reviewed (changes requested):
Thanks for your work on this new feature @bejancsaba! I noted a number of small recommendations related to logging and error messages. I also raised a question about timestamp conversion, which should not be happening in current versions.

bejancsaba commented on nifi-nar-bundles/nifi-gcp-bundle/nifi-gcp-processors/src/main/java/org/apache/nifi/processors/gcp/bigquery/PutBigQuery.java:
```diff
@@ -382,12 +368,11 @@ private WriteStream createWriteStream() {
-            getLogger().error("Failed to create Big Query Write Client for writing due to {}", new Object[] {e});
-            setupException.set(e);
+        } catch (Exception e) {
+            throw new ProcessException("Failed to create Big Query Write Client for writing due to", e);
```
Unfortunately I was not able to find a better (NiFi defined) one. There is a processor called ProcessorInstantiationException but that is defined on the controller package and is a checked one not Runtime so doesn't really fit this use case. Other than that I haven't seen anything that would be better.

bejancsaba commented on nifi-nar-bundles/nifi-gcp-bundle/nifi-gcp-processors/src/main/java/org/apache/nifi/processors/gcp/bigquery/PutBigQuery.java:
```diff
@@ -0,0 +1,457 @@
+    }
+
+    @Override
+    public void onTrigger(ProcessContext context, ProcessSession session) throws ProcessException {
```
The AbstractProcessor abstract class defines the onTrigger with this throws clause at Overriding we are just keeping that signature.

arpadboda replied:
Okay, thanks!

bejancsaba commented:
@exceptionfactory thanks for your review I addressed all your comments except one where I added some context / explanation when you have some time please take a look.

exceptionfactory reviewed (approved):
Thanks for making the updates @bejancsaba, the latest version looks good. I will defer to @arpadboda for final review.
